crates/wasmtime/src/linker.rs

```diff
@@ -270,6 +271,117 @@ impl Linker {
+    /// exports to the `Linker` as the `instance` method.
+    ///
+    /// For a Command module, this adds exports to the `Linker` which
+    /// instantiate the module each time the exports are called.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the any item is redefined twice in this linker (for
+    /// example the same `module_name` was already defined) and shadowing is
+    /// disallowed, or if `instance` comes from a different [`Store`] than this
+    /// [`Linker`] originally was created with.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use wasmtime::*;
+    /// # fn main() -> anyhow::Result<()> {
+    /// # let store = Store::default();
+    /// let mut linker = Linker::new(&store);
+    ///
+    /// // Instantiate a small instance and inform the linker that the name of
+    /// // this instance is `instance1`. This defines the `instance1::run` name
+    /// // for our next module to use.
+    /// let wat = r#"(module (func (export "run") ))"#;
+    /// let module = Module::new(&store, wat)?;
+    /// linker.module("instance1", &module)?;
+    ///
+    /// let wat = r#"
+    ///     (module
+    ///         (import "instance1" "run" (func $instance1_run))
+    ///         (func (export "run")
+    ///             call $instance1_run
+    ///         )
+    ///     )
+    /// "#;
+    /// let module = Module::new(&store, wat)?;
+    /// let instance = linker.instantiate(&module)?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn module(&mut self, module_name: &str, module: &Module) -> Result<&mut Self> {
+        match ModuleKind::categorize(module)? {
+            ModuleKind::Command => self.command(module_name, module),
+            ModuleKind::Reactor => {
+                let instance = self.instantiate(&module)?.start()?;
+
+                if let Some(export) = instance.get_export("_initialize") {
+                    if let Extern::Func(func) = export {
+                        func.get0::<()>()
+                            .and_then(|f| f().map_err(Into::into))
+                            .context("calling the Reactor initialization function")?;
+                    }
+                }
+
+                self.instance(module_name, &instance)
+            }
+        }
+    }
+
+    fn command(&mut self, module_name: &str, module: &Module) -> Result<&mut Self> {
+        for export in module.exports() {
+            if let Some(func_ty) = export.ty().func() {
+                let imports = self.compute_imports(module)?;
+                let module = module.clone();
+                let export_name = export.name().to_owned();
+                let func = Func::new(&self.store, func_ty.clone(), move |_, params, results| {
+                    let instance = Instance::new(&module, &imports)
+                        .map_err(|error| match error.downcast::<Trap>() {
+                            Ok(trap) => trap,
+                            Err(error) => Trap::new(error.to_string()),
```

Comment:
This I think may want to use the debug formatter instead of `Display` to capture the full context of the `anyhow::Error`